src/route/api.ts

```diff
@@ -1,23 +1,87 @@
 import { Request, Response, Router } from 'express';
 import multer from 'multer';
+import pino from 'pino';
 
-import { processCSV } from '../controllers/csv-processor';
+import { processCSV, uploadCSV } from '../controllers/csv-processor';
+import { listFilesInDirectory, downloadFileFromDataLake } from '../controllers/datalake';
 
 const storage = multer.memoryStorage();
 const upload = multer({ storage });
 
+export const logger = pino({
+    name: 'StatsWales-Alpha-App',
+    level: 'debug'
+});
+
 export const apiRoute = Router();
 
+logger.debug('API is runnuing...');
+
 apiRoute.get('/', (req, res) => {
     res.json({ message: 'API is available' });
 });
 
 apiRoute.post('/csv', upload.single('csv'), async (req: Request, res: Response) => {
+    // const page_number_str: string = req.query.page_number || req.body?.page_number;
+    // const page_size_str: string = req.query.page_size || req.body?.page_size;
+    // const page_number: number = Number.parseInt(page_number_str, 10) || 1;
+    // const page_size: number = Number.parseInt(page_size_str, 10) || 100;
+    const filemame: string | undefined = req.query.filename?.toString() || req.file?.originalname;
+    if (filemame === undefined) {
+        res.status(400);
+        res.json({
+            success: false,
+            headers: undefined,
+            data: undefined,
+            errors: [
+                {
+                    field: 'filename',
+                    message: 'No filename provided'
+                }
+            ]
+        });
+        return;
+    }
+    const processedCSV = await uploadCSV(req.file?.buffer, filemame);
+    if (!processedCSV.success) {
+        res.status(400);
+    }
+    res.json(processedCSV);
+});
+
+apiRoute.get('/csv/', async (req, res) => {
+    const fileList = await listFilesInDirectory();
+    res.json({ filelist: fileList });
+});
+
+apiRoute.get('/csv/:file', async (req, res) => {
+    const filename = req.params.file;
+    const file = await downloadFileFromDataLake(filename);
+    if (file === undefined || file === null) {
+        res.status(404);
+        res.json({ message: 'File not found... file is null or undefined' });
+        return;
+    }
+    res.setHeader('Content-Length', file.length);
+    res.setHeader('Content-Type', 'text/csv');
+    res.setHeader('Content-Disposition', `attachment; filename=${filename}`);
+    res.write(file, 'binary');
+    res.end();
+});
+
+apiRoute.get('/csv/:file/view', async (req, res) => {
+    const filename = req.params.file;
+    const file = await downloadFileFromDataLake(filename);
```

Comment:
We read the entire file in every time, even if we only want a small chunk of it.
If we can't avoid that, is there at least a way we can process each chunk as it comes in and only store it if it's part of the page we're looking for? This way we only need 1 chunk of memory plus enough for the size of the page. Otherwise the whole file ends up in core for the whole duration of the Request/Response and most (100% - (page_size/file_size)) of it ends up as garbage that needs to be collected.
Ironically, the smaller the pages (i.e. the more we decrease the burden on the client), the more memory we end up wasting on the server as a percentage of chunks transmitted to the client.

Reply:
I agree with this... once we move towards having metadata and a database this is exactly what I'd like to do.